### PR TITLE
Add realm secret rotation.

### PR DESCRIPTION
**Pull request by MancunianSam**
I've added in a class and some tests to rotate the realm keys. I've also
refactored the way it sends messages to the notifications service which
is linked to
https://github.com/nationalarchives/tdr-notifications/pull/81.

### CODE REVIEW

File: src/main/scala/uk/gov/nationalarchives/rotate/RotateRealmKeys.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+class RotateRealmKeys(client: Keycloak) {
+  val logger: Logger = new SimpleLoggerFactory().getLogger(this.getClass.getName)
+
+  def rotate(): Message = Try {
+    val realm = client.realm("tdr")
+    val components: ComponentsResource = realm.components()
+    val keyProvider = "org.keycloak.keys.KeyProvider"
+    val keyName = "rsa-generated"
+
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/rotate/RotateRealmKeys.scala:22]: Unnecessary line break.

File: src/main/scala/uk/gov/nationalarchives/rotate/RotateRealmKeys.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    val rsaKeys = components.query("tdr", keyProvider, keyName)
+
+    def filterKeys(filterValue: String): List[ComponentRepresentation] = rsaKeys.asScala.filter(key => key.getConfig.get("active").asScala.toList.head == filterValue).toList
+
+    val inactiveKeys: List[ComponentRepresentation] = filterKeys("false")
+    val activeKeys: List[ComponentRepresentation] = filterKeys("true")
+    //Delete the passive keys from last week, no-one should be using them.
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/rotate/RotateRealmKeys.scala:30]: Maybe reword this comment, in case the schedule for rotating changes: `Delete the previous passive keys, no-one should be using them`

File: src/test/scala/uk/gov/nationalarchives/rotate/RotateRealmKeysSpec.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    .realm("tdr")
+    .authorization("token")
+    .build()
+
+  def loggedRequest(method: RequestMethod): LoggedRequest =
+    authServer.getAllServeEvents.asScala
+      .find(_.getRequest.getMethod == method)
+      .map(_.getRequest)
+      .get
+
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/rotate/RotateRealmKeysSpec.scala:76]: Unnecessary line break

File: src/test/scala/uk/gov/nationalarchives/rotate/RotateRealmKeysSpec.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  "The rotate method" should "should invalidate the active key" in {
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/rotate/RotateRealmKeysSpec.scala:78]: No need for the word "should" in the test title

**Review by TomJKing — COMMENTED** [review]
Looks good. Thanks. Just some minor comments. The README will need updating as well.

Also is rotation also required for the master realm?  I don't think it is.

**Comment by MancunianSam** [comment]
The realm keys are for signing tokens which we don't use on the master realm so we don't need to rotate them there. I've updated the README with a description of what the key rotation does as well.

**Review by TomJKing — APPROVED** [review]
Looks good. Thanks
